### PR TITLE
[v5] `onSnapshot` for all actors

### PR DESCRIPTION
**Pull request by davidkpiano**
This PR adds `onSnapshot: { ... }` support for all supported invoked actor logic.

### CODE REVIEW

**Comment by davidkpiano** [comment]
@Andarist I need help because the test failures are only related to `console.warn` unexpected calls (locally, `yarn test` passes ✅ )

File: packages/core/src/StateMachine.ts [hunk 1]
```diff
@@ -384,10 +384,10 @@ export class StateMachine<
     state: State<TContext, TEvent, TActor, TOutput, TResolvedTypesMeta>
   ) {
     return state.error
-      ? { status: 'error', data: state.error }
+      ? ({ status: 'error', error: state.error } as const)
       : state.done
-      ? { status: 'done', data: state.output }
-      : { status: 'active' };
+      ? ({ status: 'done', data: state.output } as const)
+      : ({ status: 'active' } as const);
```
Comment (Andarist) [inline packages/core/src/StateMachine.ts:390]: this is unrelated change

File: packages/core/src/actions/invoke.ts [hunk 2]
```diff
@@ -123,6 +144,7 @@ export function invoke<
   invoke.systemId = systemId;
   invoke.src = src;
   invoke.input = input;
+  invoke.subscribe = !!onSnapshot;
```
Comment (Andarist) [inline packages/core/src/actions/invoke.ts:147]: I like this one very much - but perhaps we could rename this somehow? `syncSnapshot`? `iLikeChatter`? `listenToSnapshots`?

File: packages/core/src/actions/invoke.ts [hunk 3]
```diff
@@ -53,6 +55,23 @@ function resolve(
             })
           : configuredInput
     });
+
+    if (subscribe) {
+      actorRef.subscribe({
+        next: (snapshot) => {
+          if (actorContext.self.status === ActorStatus.Running) {
```
Comment (Andarist) [inline packages/core/src/actions/invoke.ts:62]: is there any test case that would exercise the need for this if statement?

File: packages/core/src/actions/invoke.ts [hunk 4]
```diff
@@ -53,6 +55,23 @@ function resolve(
...
+            actorContext.self.send({
+              type: `xstate.snapshot.${id}`,
+              id,
+              snapshot
+            });
+          }
+        },
+        error: () => {
+          /* TODO */
+        }
```
Comment (Andarist) [inline packages/core/src/actions/invoke.ts:72]: that shouldn't be here as if we are going to add an error listener to the `actorRef` then it will be done outside of this if block

File: packages/core/src/spawn.ts [hunk 5]
```diff
@@ -45,16 +45,52 @@ export function createSpawner(
             : input,
         systemId
       }) as any;
-      spawnedChildren[actor.id] = actor;
-      return actor;
+      spawnedChildren[actorRef.id] = actorRef;
+
+      // if (options.subscribe) {
```
Comment (Andarist) [inline packages/core/src/spawn.ts:50]: looks like a leftover - this if statement here should be uncommented

File: packages/core/src/spawn.ts [hunk 6]
```diff
@@ -45,16 +45,52 @@ export function createSpawner(
...
+            actorContext.self.send({
+              type: `xstate.snapshot.${actorRef.id}`,
+              snapshot,
+              id: actorRef.id
+            });
+          }
+        },
+        error: () => {
+          /* TODO */
+        }
```
Comment (Andarist) [inline packages/core/src/spawn.ts:63]: same as https://github.com/statelyai/xstate/pull/4172/files#r1304354021

File: packages/core/src/spawn.ts [hunk 7]
```diff
@@ -45,16 +45,52 @@ export function createSpawner(
...
       // TODO: this should also receive `src`
-      return createActor(src, {
+      const actorRef = createActor(src, {
         id: options.id,
         parent: actorContext.self,
         input: options.input,
         systemId
       });
+
+      // if (options.subscribe) {
```
Comment (Andarist) [inline packages/core/src/spawn.ts:76]: same as above

File: packages/core/src/spawn.ts [hunk 8]
```diff
@@ -45,16 +45,52 @@ export function createSpawner(
...
+            actorContext.self.send({
+              type: `xstate.snapshot.${actorRef.id}`,
+              snapshot,
+              id: actorRef.id
+            });
+          }
+        },
+        error: () => {
+          /* TODO */
+        }
```
Comment (Andarist) [inline packages/core/src/spawn.ts:89]: same as above

File: packages/core/test/invoke.test.ts [hunk 9]
```diff
@@ -2048,20 +2069,23 @@ describe('invoke', () => {
       actor.subscribe({
         complete: () => {
           done();
-        }
+        },
+        error: () => {}
```
Comment (Andarist) [inline packages/core/test/invoke.test.ts:2073]: this looks like an unrelated change - and one that could suppress legit errors in the future

File: packages/core/src/actions/invoke.ts (hunk 2 above)
Comment (davidkpiano) [reply packages/core/src/actions/invoke.ts:147]: `syncSnapshot` it is

File: packages/core/src/actions/invoke.ts (hunk 4 above)
Comment (davidkpiano) [reply packages/core/src/actions/invoke.ts:72]: Not sure how to resolve this; removing this causes a test failure

File: packages/core/src/interpreter.ts [hunk 10]
```diff
@@ -225,8 +225,8 @@ export class Actor<
         break;
       case 'error':
         this._stopProcedure();
-        this._error(status.data);
-        this._parent?.send(error(this.id, status.data));
+        this._parent?.send(error(this.id, status.error));
+        this._error(status.error);
```
[the remark this answers is not in the document]
Comment (davidkpiano) [reply packages/core/src/interpreter.ts:229]: I don't remember so I un-reordered it

File: .changeset/mean-pianos-explode.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
+---
+'xstate': minor
+---
+
+The `onSnapshot: { ... }` transition object is now supported for invoked machines, observables, promises, and transition functions:
+
+```ts
+const machine = Machine({
```
Comment (Andarist) [inline .changeset/mean-pianos-explode.md:8]:
```suggestion
const machine = createMachine({
```

File: .changeset/mean-pianos-explode.md [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+      }
+    },
+    {
+      src: fromPromise(() => { ... }),
+      onSnapshot: {
+        actions: (context, event) => {
+          event.snapshot; // promise value
+        }
+      }
+    },
```
Comment (Andarist) [inline .changeset/mean-pianos-explode.md:26]:
```suggestion
```

File: packages/core/src/actions/invoke.ts (hunk 4 above)
Comment (Andarist) [reply packages/core/src/actions/invoke.ts:72]: I'm not 100% sure what I was thinking about here 😅 I had some concerns related to this right now but I think they are addressed, see this test case:
https://github.com/statelyai/xstate/blob/23402e34c178ba9a000e22c54765090b8fcfc52f/packages/core/test/errors.test.ts#L213

File: packages/core/test/actor.test.ts [hunk 13]
```diff
@@ -487,7 +491,10 @@ describe('spawning observables', () => {
         idle: {
           entry: assign({
             observableRef: ({ spawn }) => {
-              const ref = spawn(observableLogic, { id: 'int' });
+              const ref = spawn(observableLogic, {
+                id: 'int',
+                syncSnapshot: true
```
Comment (Andarist) [inline packages/core/test/actor.test.ts:496]: It's worth noting that we don't have any means right now to validate that this is actually safe on the type-level

**Review by Andarist — APPROVED** [review]
